src/nxt_main_process.c

```diff
@@ -1091,131 +1092,6 @@ nxt_main_port_socket_handler(nxt_task_t *task, nxt_port_recv_msg_t *msg)
 }
 
 
-static nxt_int_t
-nxt_main_listening_socket(nxt_sockaddr_t *sa, nxt_listening_socket_t *ls)
```

Comment:
It seems you didn't notice that the `router` process is different from the `controller` process.
`controller` has the privilege to do `bind` on the controller listening socket, then it does `listen` operation.
But `router has no such `privilege`, so it has to ask `main` process to do `bind` on listening sockets from the configuration. The `main` process doesn't do `listen`, the `router` process does.

There are two operations: `bind` and `listen`.
`controller` process: `bind` + `listen`
`router` process: ask `main` to `bind` then `listen` by itself.

So it's wrong to reuse `nxt_listen_socket_create()`.

Reply:
Sorry, didn't really notice that difference.

I asked @VBart earlier about the need to keep the separation between nxt_listen_socket and nxt_main_listening_socket.

My question and his answer:
https://github.com/nginx/unit/pull/649#issuecomment-1044377567

Reply:
listening socket in `controller` process is in charge of configuration management.
listening in `router` process is in charge of traffic requests.
So the `listen` operation related to `listeners` options need to be done in the `router` process.

Reply:
Thanks for your explanation! 

The sequence and reasons for calling via RPC between `main <-> router` are clear to me.

Please explain. I don't understand why `nxt_listen_socket` is executed exactly in `router` in `nxt_router_listen_socket_ready`? After all, in fact, this happens immediately after the creation of the socket and after the response from `main -> router`. If an error occurs in the main RPC handler, then the `router` will simply not receive `fd` from `main` and will not interact with the given socket. From the point of view of parameters: `BACKLOG`, `TCP_DEFER_ACCEPT` and other sockets for the `controller` and for the router seem identical to me.

**Both cases:**
- `controller` -> `nxt_listen_socket_create`
- `router` -> `main(nxt_main_listening_socket)` -> `router(nxt_router_listen_socket_ready)`

The same socket setup happens, only in the case of the `router` it is split into two parts and ends in `nxt_router_listen_socket_ready`. Is it only necessary to put the socket in passive mode only on successful completion of the RPC exchange between processes in the `router`? Does it make any real sense?

Reply:
If so, then I suggest keeping the current split and just adding unlink before bind in `nxt_main_listening_socket`, thus solving this problem in the same way as it is already implemented in `nxt_listen_socket_create`. And in the future, we can replace unlink with a more secure solution, if at the moment `nxt_socket_release_by_path` does not suit us for one reason or another.

Reply:
Because the `router` process is designed to process traffic/http requests.
It's possible that some requests might cause the host process (here is `router`) to crash in the real world. If it happens in the `router ` process, the `main` process is able to create a new one. That's why the `main` process doesn't process traffic requests.

Reply:
If you understand why `router` process is in charge of traffic requests, you would know the `listen` operation should do in `router` process. But all the listening socket should be bonded first, it need to do by `main` process.
After all the sockets are bonded successfully, the `router` process start `listen` them.
Note that `router` process is multiple-threads, all the threads have the same chance to get notifications on the same listening socket because `epoll` is thread-safe.
Yeah, Unit is a more complicated oss, devs need to be more patient if they are interested in its source code.

Reply:
As far as I know, listen has an effect on the socket itself and is not tied to the state of a particular process. In my understanding, listen simply switches the socket to passive mode, setting the mode of operation for incoming connections to be processed when accept is received. So I can't see a direct relationship between the traffic (which is the function of the router) and the process (which is the router) with the listen call itself, which simply changes the socket behavior.

Reply:
> If so, then I suggest keeping the current split and just adding `unlink` before `bind` in `nxt_main_listening_socket`, thus solving this problem in the same way as it is already implemented in `nxt_listen_socket_create`. And in the future, we can replace `unlink` with a more secure solution, if at the moment `nxt_socket_release_by_path` does not suit us for one reason or another.

I agree to use `unlink` now, but we need to reuse the logic in `nxt_listen_socket_create()`: create temp file then bind.
Otherwise, if it's bonded with other instances like nginx, it's bad to `unlink` it.

Reply:
I understand that according to the logic of the device (architecturally), it is logical to activate incoming connections in the router, since it controls the flow of traffic. But purely technically and in fact, at the moment there is practically no difference when activating listen in the router or in main. I'm trying to see the real benefits of listen in nxt_router_listen_socket_ready except that it seems logical.

Reply:
> As far as I know, listen has an effect on the socket itself and is not tied to the state of a particular process. In my understanding, listen simply switches the socket to passive mode, setting the mode of operation for incoming connections to be processed when accept is received. So I can't see a direct relationship between the traffic (which is the function of the router) and the process (which is the router) with the listen call itself, which simply changes the socket behavior.

You are right, `listen` only changes the bound TCP socket into a listening state.

Reply:
> > If so, then I suggest keeping the current split and just adding `unlink` before `bind` in `nxt_main_listening_socket`, thus solving this problem in the same way as it is already implemented in `nxt_listen_socket_create`. And in the future, we can replace `unlink` with a more secure solution, if at the moment `nxt_socket_release_by_path` does not suit us for one reason or another.
> 
> I agree to use `unlink` now, but we need to reuse the logic in `nxt_listen_socket_create()`: create temp file then bind. Otherwise, if it's bonded with other instances like nginx, it's bad to `unlink` it.

Then we need to split `nxt_listen_socket_create` into code before and after `listen` execution, right?

Reply:
The logic of "create temp path for unix domain socket" and "test connection after listening" should work in the same process. If `bind` happens on the `main` process but `listen` happens on the `router` process, the logic is broken.
We'll discuss it first and tell you if it's clear.

Reply:
Now I don't understand at all. So this is the current version of the existing code in the origin master branch. Do you think it's incorrect?

Reply:
I think it's right. It makes sense that the router process does the `listen` operation until all the listeners have been bounded in the main process.
What I'm not sure about is if the `listen` operation can be moved to the `main` process from the `router` process. I didn't notice the background before.